tests/test_voxelmorph.py

```diff
@@ -0,0 +1,316 @@
+    # only one level in the UNet
+    {
+        "spatial_dims": 3,
+        "in_channels": 2,
+        "unet_out_channels": 32,
+        "channels": (16, 32),
+        "final_conv_channels": (16, 16),
+    },
+    ((2, 1, 160, 192, 224), (2, 1, 160, 192, 224)),
+    ((2, 1, 160, 192, 224), (2, 3, 160, 192, 224)),
+]
+
+CASES = [
+    TEST_CASE_0,
+    TEST_CASE_1,
+    TEST_CASE_2,
+    TEST_CASE_3,
+    TEST_CASE_4,
+    TEST_CASE_5,
+    TEST_CASE_6,
+    TEST_CASE_7,
+    TEST_CASE_8,
+    TEST_CASE_9,
+    TEST_CASE_10,
+    TEST_CASE_11,
+]
+
+ILL_CASE_0 = [  # spatial_dims = 1
+    {
+        "spatial_dims": 1,
+        "in_channels": 2,
+        "unet_out_channels": 32,
+        "channels": (16, 32, 32, 32, 32, 32),
+        "final_conv_channels": (16, 16),
+    }
+]
+
+ILL_CASE_1 = [  # in_channels = 3 (not divisible by 2)
+    {
+        "spatial_dims": 3,
+        "in_channels": 3,
+        "unet_out_channels": 32,
+        "channels": (16, 32, 32, 32, 32, 32),
+        "final_conv_channels": (16, 16),
+    }
+]
+
+ILL_CASE_2 = [  # len(channels) = 0
+    {
+        "spatial_dims": 3,
+        "in_channels": 2,
+        "unet_out_channels": 32,
+        "channels": (),
+        "final_conv_channels": (16, 16),
+    }
+]
+
+ILL_CASE_3 = [  # channels not in pairs
+    {
+        "spatial_dims": 3,
+        "in_channels": 2,
+        "unet_out_channels": 32,
+        "channels": (16, 32, 32, 32, 32, 32, 32),
+        "final_conv_channels": (16, 16),
+    }
+]
+
+ILL_CASE_4 = [  # len(kernel_size) = 3, spatial_dims = 2
+    {
+        "spatial_dims": 2,
+        "in_channels": 2,
+        "unet_out_channels": 32,
+        "channels": (16, 32, 32, 32, 32, 32, 32),
+        "final_conv_channels": (16, 16),
+        "kernel_size": (3, 3, 3),
+    }
+]
+
+ILL_CASE_5 = [  # len(up_kernel_size) = 2, spatial_dims = 3
+    {
+        "spatial_dims": 3,
+        "in_channels": 2,
+        "unet_out_channels": 32,
+        "channels": (16, 32, 32, 32, 32, 32, 32),
+        "final_conv_channels": (16, 16),
+        "up_kernel_size": (3, 3),
+    }
+]
+
+ILL_CASES = [ILL_CASE_0, ILL_CASE_1, ILL_CASE_2, ILL_CASE_3, ILL_CASE_4, ILL_CASE_5]
+
+
+class TestVOXELMORPH(unittest.TestCase):
+    @parameterized.expand(CASES)
+    def test_shape(self, input_param, input_shape, expected_shape):
+        net = VoxelMorph(**input_param).to(device)
+        with eval_mode(net):
+            result = net.forward(torch.randn(input_shape[0]).to(device), torch.randn(input_shape[1]).to(device))
+            self.assertEqual(result[0].shape, expected_shape[0])
+            self.assertEqual(result[1].shape, expected_shape[1])
+
+    def test_script(self):
+        net = VoxelMorph(
+            spatial_dims=3,
+            in_channels=2,
+            unet_out_channels=32,
+            channels=(16, 32, 32, 32, 32, 32),
+            final_conv_channels=(16, 16),
+        ).net
+        test_data = torch.randn(1, 2, 160, 192, 224)
+        test_script_save(net, test_data)
```

Comment:
Hi @wyli, I wonder if this would be the appropriate way to test saving the script. I tried to save the entire model (including the `DVF2DDF` for velocity field integration and `Warp` for applying the deformation). But it looks like jit doesn't like the way `Warp` is implemented:

```
RuntimeError: 
cannot statically infer the expected size of a list in this context:
  File "/home/runner/work/MONAI/MONAI/monai/networks/blocks/warp.py", line 125
        if spatial_dims not in (2, 3):
            raise NotImplementedError(f"got unsupported spatial_dims={spatial_dims}, currently support 2 or 3.")
        ddf_shape = (image.shape[0], spatial_dims) + tuple(image.shape[2:])
                                                     ~~~~~~~~~~~~~~~~~~~~ <--- HERE
```

So, for the sake of passing the tests, I am trying to save only the network part. Is there a more elegant way of passing the tests without having to leave these two modules behind? Or would it be possible to just rewrite this line in a way that jit likes?

https://github.com/Project-MONAI/MONAI/blob/cf886e7dc41361b61df3d612d897963d4043c61d/monai/networks/blocks/warp.py#L125

Reply:
thanks for looking into this, it seems that the `ddf_shape` variable is just evaluated for a minor shape checking https://github.com/Project-MONAI/MONAI/blob/cf886e7dc41361b61df3d612d897963d4043c61d/monai/networks/blocks/warp.py#L125-L130 I feel we can just remove the check in favour of jit.